### PR TITLE
Adding Fletching Factory and Spectral Arrows. Also, issues

### PR DESCRIPTION
**Pull request by Akiyama64**
Fletching factory added. To get it, you upgrade from a laboratory with 512 arrows, 256 eggs, 128 hay bales, 64 nether warts, 32 oak saplings, 32 spruce saplings, and 4 water buckets (which turn into buckets after running the factory). 

The reasoning behind this recipe is that these things all have sticky combinations, and nether wart is useful for potions anyway. Also, the materials come from different shards.

The Spectral Arrows come from inputting 16 arrows and 16 glowstone dust to create 64 spectral arrows. In vanilla minecraft, you can create 32 spectral arrows using just 16 arrows and a 64 glowstone dust. The factory recipe is more efficient by doubling output and cutting glowstone to a quarter of the cost. 

Now for the issues. There does not appear to be a way to create potion tipped arrows with factory mod. In fact, Factory Mod does not seem to recognize potions either. That means that inputs and outputs seem to not have a way to understand the input and the output. If durability can help the factory understand potions and potion tipped arrows (like it does with dyes and saplings) in the future (if not already done so), then we can make some work on the potion arrows.

(For later recipe reference, one lingering potion can create 8 potion arrows in vanilla Minecraft. I recommend that the factory produce more arrows than that. Like 16.)

Anyhow, this should be a good start for creating special arrows. Let us know how we can make the factories understand potion tipped arrows and potions and things can move forward.

### CODE REVIEW

**Comment by Maxopoly** [comment]
Upgrades dont support outputs. Is there any reasoning behind the upgrade cost?

> 256 eggs, 128 hay bales, 64 nether warts, 32 oak saplings, 32 spruce saplings, and 4 water buckets

It seems kinda random (and literally free). There's no way to get the arrows, so vanilla isn't really an argument imo, we can just keep it in line with the crafting recipe or do something else if we want. How are those balance wise, arent some of those just health pots going to a stack of 64? Do we need to nerf/disable anything before putting those potion arrows online?

Also allowing them to be parsed from the config isn't too hard, I can do that later. Potion data used to be just durability, but they changed it in 1.9.

**Comment by Akiyama64** [comment]
The reasoning behind the initial costs were how i was thinking of sticky materials, or things that are sticky in combination. Wheat's gluten, egg yolks, tree sap, and netherwart to give it all a sticky mess of goo once the water mixes in. Keep in mind that the fletching factory is an upgrade from the laboratory, so there are more costs to making one.

Still, since you say the costs are too low, how about requiring 64 of every sapling, 128 nether wart, and 256 hay bales? That's a lot of tree sap (which is a the tree's glue like blood), and a lot more gluten. 

Or, perhaps a different factory upgrade recipe. I was thinking of a factory made to stick things together, so what other things do we have for that? Perhaps an entirely different approach to the factory is better though. (Slime would be good, for example, but it does not exist.)

As for the vanilla part, you have a point. I think making 8 potion arrows from 1 potion is legit for future balance (it's vanilla anyway).

As long as strength II does not sneak in, it should be okay. However, I'm not much of a pvper, so asking about how OP poison arrows may be on the subreddit is recommended. (I can do that for you for potion arrow balance, if you like.)

**Comment by Maxopoly** [comment]
Yeah, finding people who actually know stuff about 1.9 might be a bit hard, but would be appreciated to make sure nothing there is completly broken.

**Comment by ProgrammerDan** [comment]
As for parsing / outputting a tipped arrow, unless we want to do it a different way, you can specify the NBT data directly using the NBT extensions I added to CivModCore. I can whip up a sample later if that's useful.

Similarly Spawn Eggs will need NBT data, unless Spigot finally got around to finishing that api...

**Comment by Akiyama64** [comment]
NBT data? I wonder what that will look like, and what does NBT mean?

Nevertheless, it sounds useful.

And now I'll get to increasing the costs like I said.

**Comment by Maxopoly** [comment]
All potion items have their own itemmeta class, so I'll use that instead, because it's easier to add new stuff and cleaner code wise.
https://hub.spigotmc.org/javadocs/spigot/org/bukkit/inventory/meta/PotionMeta.html


But yeah for spawn eggs we'll definitely need your nbt thing.

**Comment by ProgrammerDan** [comment]
It's how minecraft stores everything, it's kind of a custom YAML / JSON block attached to an item.

@Maxopoly Hence my "finish the api" comment. When I first updated to 1.9, that API was empty. It literally did nothing, just wrapped the old code without injecting the new NBT tags. That's why ArthropodEggs now uses Spigot instead of spigot-api ...

**Comment by Maxopoly** [comment]
Wait so that PotionMeta thing isn't working? Also that might explain why people said ArthropodEgg on civ is no longer working.

**Comment by ProgrammerDan** [comment]
Ah yes w.r.t. metas -- can we please move towards just using ConfigurationSerialized segments? 

Check here for an example: https://github.com/Civcraft/SimpleAdminHacks/blob/master/src/main/resources/config.yml#L40

Basically lets us use all the Spigot-backed structures without any effort. No effort at all really ...

**Comment by ProgrammerDan** [comment]
w.r.t ArthropodEgg just merge in Devoted's v1.9 branch.

**Comment by Maxopoly** [comment]
Hmm, the serialization stuff is kinda okay, but I have yet to find documentation on it (especially for the more complicated meta stuff), so having our own thing with docu seems better and allows others to make PRs. Also it doesn't allow some of the nice stuff needed for ItemMaps like amount far bigger than stack sizes or wildcarding (which I want to extend to enchants, lore and meta in general).

**Comment by ProgrammerDan** [comment]
Fair point, just annoys me that we're recreating the wheel all the time. I'd rather do this: Specify Base Type using spigot's serialization, and have overrides that allow expanding on it (no subtype restriction, stack size override, etc.)

That way we use the core functionality as is without recoding it, and only have to actually implement our special sauce instead of always re-implementing everything.

**Comment by Maxopoly** [comment]
Hmm, how much effort would it be to make a script to convert the current config over?

**Comment by ProgrammerDan** [comment]
Not too much effort I'd expect, since the config resolves to ItemStacks anyway.

**Comment by ProgrammerDan** [comment]
I'll whip up some examples of serialized types for each major Meta grouping, that should make it a bit more accessible.

**Comment by Maxopoly** [comment]
Hmm, I guess if we converted over the whole config we could do that, but then we would also have to write extensive documentation on the serialization format and at that point I'm not sure whether it's worth the effort.

I mean once we have the potion stuff, the only thing we can't natively do are written books, fireworks and some other specific meta stuff which is really easy to add, if we ever want it. Basically we are already done reinventing the wheel.

**Comment by ProgrammerDan** [comment]
True. 'tis a tragedy, but largely past.

tbh it's a discussion we should have had 6-7 months ago, but that's mostly my fault. i was still experimenting with them, then; now I use them exclusively in my development so dealing with Civcraft's mainline approach feels like a regression to me, but I can see how maintaining the momentum and useage experience makes sense.

**Comment by Akiyama64** [comment]
In addition to the recent cost changes that I've made to the upgrade cost, how would adding 16 ghast tears to the upgrade cost work for you all? That way, we can totally get rid of the water bucket cost and use tears to mix everything together. [As DarkJesus suggested,](https://www.reddit.com/r/Civcraft/comments/4il35t/serious_how_should_potion_arrows_be_balanced_if/d2yxm7q?context=3) this means players have to hunt for their goods.

In summary, the upgrade cost from the lab would be 512 arrows; 256 eggs, netherwart, and hay bales; 64 of every sapling; and 16 ghast tears.

**Comment by Crushtoe** [comment]
I'd suggest not adding Healing arrows since it's possible to turn into stackable healing pots.

Regen arrows are also easily abused; shoot someone with a regen pot active and  it resets their effect duration from 8 minutes to 2 seconds. You could also use it to apply quick regen. It would just be too easily abused.


Arrows are an essential part of PvP; they are bottleneck end by their cost and the low quantities they drop in. A factory should have a decent cost to it, like 576 arrows (9 stacks) if it's going to be making tons of cheap arrows.

**Comment by Akiyama64** [comment]
Absolutely agree with that. I learned of that through this thread where I asked about balance: https://www.reddit.com/r/Civcraft/comments/4il35t/serious_how_should_potion_arrows_be_balanced_if/

Crushtoe, it's likely that the devs will force all special arrows to be unstackable. Just like pots, there can only be one arrow per slot. Otherwise, they will not be added because of how unfair they are in stacks.

Note to Max and Dan, the[ balancing discussion thread](https://www.reddit.com/r/Civcraft/comments/4il35t/serious_how_should_potion_arrows_be_balanced_if/) has lots of talk going on that addresses balancing issues.

I believe that the factory upgrade cost that I proposed today would be best as a LOW TIER fletching factory that produces arrows that are more for fun than combat. The high tier one has not been designed yet.

**Comment by Maxopoly** [comment]
@Akiyama64 https://github.com/Civcraft/FactoryMod/blob/master/config.yml#L74

Here's how to specify potions in the config

**Comment by Akiyama64** [comment]
Thanks. I've taken your advice on how to change the upgrade costs. Also, a jump arrow recipe is in it now, but there are many more to be added over time.

**Comment by Crushtoe** [comment]
Have you tried adding projectile projection to armor? I imagine even Projectile Projection III on your boots would make healing arrows useful, and Prot IV only actually gives you any bonus damage reduction up to 3 armor pieces.

**Comment by Akiyama64** [comment]
Just checked it out. Here's a video to show it: https://youtu.be/3SZ6w9QS4R4

Health potion arrows are garbage in vanilla Minecraft. They are so useless that nobody would make them anyway, no matter how easy.

Also, Max, how do you feel about making two tiers of a fletching factory as Crushtoe suggested?

**Comment by Maxopoly** [comment]
Two tiers sounds fine to me. And yeah, it seems like we need to rebalance the arrows from scratch, which we could do with the custom effects that can be applied to the arrows.

**Comment by Akiyama64** [comment]
Wait, custom effects for a health II arrow? According to the config doc, I think the max health level you can customize with that method is Level 2, and that wouldn't be enough. Perhaps there's a way to make it Level 5 or however many it take to finally be a net good for health?

**Comment by Maxopoly** [comment]
The custom effects there allow you to overwrite the default effect and has no level limit.

**Comment by Maxopoly** [comment]
Rejecting this for now, as it needs a complete rework of potion arrows first
